pymc3/distributions/distribution.py

```diff
@@ -141,40 +248,114 @@ def __new__(cls, name, *args, **kwargs):
                 "for a standalone distribution."
             )
 
-        rng = kwargs.pop("rng", None)
-
-        if rng is None:
-            rng = model.default_rng
-
         if not isinstance(name, string_types):
             raise TypeError(f"Name needs to be a string but got: {name}")
 
-        data = kwargs.pop("observed", None)
+        if rng is None:
+            rng = model.default_rng
 
-        total_size = kwargs.pop("total_size", None)
+        # Create the RV without specifying testval, because the testval may
+        # have a shape that only matches after replicating with a size implied
+        # by dims (see below).
+        rv_out = cls.dist(*args, rng=rng, testval=None, **kwargs)
 
-        dims = kwargs.pop("dims", None)
+        # `dims` are only available with this API, because `.dist()` can be used
+        # without a modelcontext and dims are not yet tracked at the Aesara level.
+        if dims is not None:
+            # Infer size from dims (and coords)
+            if Ellipsis in dims:
+                raise NotImplementedError("Ellipsis-based dims are not implemented.")
+                # TODO: Get implied dimensions from input Variables! (see https://github.com/pymc-devs/aesara/issues/352)
+                implied_dims = ...
+                dims = tuple(dims[:-1]) + implied_dims
 
-        if "shape" in kwargs:
-            raise DeprecationWarning("The `shape` keyword is deprecated; use `size`.")
+            dimshape = tuple(len(model.coords[dname]) for dname in dims)
+            size = tuple(dimshape[: len(dimshape) - rv_out.ndim])
```

Comment:
```suggestion
            size = tuple(len(model.coords[dname]) for dname in dims[: len(dimshape) - rv_out.ndim])
```

I think we should use this and not require all coords to be present in the model context. Conceptually we don't need them to be defined until conversion to ArviZ time, and now with size practice matches theory :tada: and they are indeed not used anymore, we get their length only to discard that info in the next line.

Reply:
Sounds good, but since we know the implied `ndim` we can also substitute the missing dim names with `f"{name}_dim_{d}"`. That should at least give us better chances of not getting downstream issues, right?

Reply:
Not sure I follow what you mean by missing dim names, we still need to pass the dim names for the implied dims right? I don't think the `: len(dimshape) - rv_out.ndim` would work otherwise. 

If we want to be extra careful, we can check that the number of dimension names provided matches the number of dimensions of the variable, but I'm not sure how we'd go about that because we can use dims to extend the implied dimensions.

If the concern is about the number of implied dimensions changing afterwards for a 2nd round of sampling, posterior predictive sampling... as long as the number of dimensions increases, ArviZ will handle that automatically and generate the `rv_name_dim_#` alone. If they are reduced it will error out, but I'd say that's on the user doing that, if they modify the number of dims they should then pass the new `dims` to ArviZ via `idata_kwargs`, `dims` or whatever. Example:

```python
import numpy as np
import arviz as az

data = np.random.normal(size=(4, 500, 5, 3, 7))
idata = az.from_dict({"a": data}, dims={"a": ["subject"]})
idata.posterior
# Dimensions:  (a_dim_1: 3, a_dim_2: 7, chain: 4, draw: 500, subject: 5)
# Coordinates:
#   * chain    (chain) int64 0 1 2 3
#   * draw     (draw) int64 0 1 2 3 4 5 6 7 8 ... 492 493 494 495 496 497 498 499
#   * subject  (subject) int64 0 1 2 3 4
#   * a_dim_1  (a_dim_1) int64 0 1 2
#   * a_dim_2  (a_dim_2) int64 0 1 2 3 4 5 6
# Data variables:
#     a        (chain, draw, subject, a_dim_1, a_dim_2) float64 0.3243 ... -0.8625
```

if you want to be cryptic and rely on default dim names you can even do:

```python
idata = az.from_dict({"a": data}, dims={"a": ["subject", "dose"]}, coords={"a_dim_2": [l for l in "abcdefg"]})
idata.posterior
# Dimensions:  (dose: 3, a_dim_2: 7, chain: 4, draw: 500, subject: 5)
# Coordinates:
#   * chain    (chain) int64 0 1 2 3
#   * draw     (draw) int64 0 1 2 3 4 5 6 7 8 ... 492 493 494 495 496 497 498 499
#   * subject  (subject) int64 0 1 2 3 4
#   * dose     (dose) int64 0 1 2
#   * a_dim_2  (a_dim_2) <U1 'a' 'b' 'c' 'd' 'e' 'f' 'g'
# Data variables:
#     a        (chain, draw, subject, dose, a_dim_2) float64 0.3243 ... -0.8625
```